### PR TITLE
Homework 8 complete

### PR DESCRIPTION
**Pull request by asish918**
# Homework 8 Asish

### Topics Covered:
- Rest Controller
- Spring Folder Structure
- @Request @Response Annotations

### Drift Video Explanation
[Drift](https://app.drift.com/video/share/abCWhxW5AP4teJuIs3dNZhiOPdBdWSDTiSdqju8Qubg)

### Output Screenshots - 
Screenshot 2024-01-25 at 00 06 38

Screenshot 2024-01-25 at 00 06 32

Screenshot 2024-01-25 at 00 06 20

Screenshot 2024-01-25 at 00 06 01

### Most of my Code Smells are because my parent package name is Spring_Exercise_3. Sonarqube expects me to use small case. I have a lot of files and each one has this one code smell. Please excuse it🙃

### CODE REVIEW

**Comment by asish918** [comment]
| Metric | Value |
| ------ | ----- |
| alert_status | <b>:white_check_mark: OK</b> |
| bugs | 0 |
| code_smells | 20 |
| reliability_rating | A (0 Bugs)<br/>Note: A being best, E being worst |
| security_rating | A (0 Vulnerabilities)<br/>Note: A being best, E being worst |
| vulnerabilities | 0 |

File: backend/homework/8/src/main/java/asish/kdu/Spring_Exercise_3/controller/AppController.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    @GetMapping("/vehicle/{name}")
+    public Vehicle findByName(@PathVariable String name) {
+        return service.getVehicleRequest(name);
+    }
+
+    @PostMapping("/vehicle")
+    @ResponseStatus(HttpStatus.CREATED)
+    public ResponseEntity<String> createVehicle(@RequestBody VehicleDto vehicleDto) {
+        service.addVehicleRequest(vehicleDto);
+        return ResponseEntity.ok("Vehicle added to Database");
```
Comment (rishav-kd) [inline backend/homework/8/src/main/java/asish/kdu/Spring_Exercise_3/controller/AppController.java:33]: Let the service function return ResponseEntity, Incase the function returns an exception the response can still be 200 ok

File: backend/homework/8/src/main/java/asish/kdu/Spring_Exercise_3/controller/AppController.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+/**
+ * Rest Controller for the Entire App
+ */
+@RestController
+@RequestMapping("/kdu")
```
Comment (rishav-kd) [inline backend/homework/8/src/main/java/asish/kdu/Spring_Exercise_3/controller/AppController.java:15]: put `/kdu/vehicle`

File: backend/homework/8/src/main/java/asish/kdu/Spring_Exercise_3/dtos/VehicleDto.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.experimental.Accessors;
+
+@Data
+@AllArgsConstructor
+@Accessors(fluent = true)
+public class VehicleDto {
+    public String name;
```
Comment (rishav-kd) [inline backend/homework/8/src/main/java/asish/kdu/Spring_Exercise_3/dtos/VehicleDto.java:11]: make sure to use @Valid

File: backend/homework/8/src/main/java/asish/kdu/Spring_Exercise_3/repository/VehicleRepository.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+@Repository
+public class VehicleRepository {
+    private List<Vehicle> db;
+    @PostConstruct
+    public void init() {
+        this.db = new ArrayList<>();
+    }
+
+    public void addVehicle(Vehicle vehicle) {
+        db.add(vehicle);
```
Comment (rishav-kd) [inline backend/homework/8/src/main/java/asish/kdu/Spring_Exercise_3/repository/VehicleRepository.java:20]: validate vehicle

Comment (rishav-kd) [reply backend/homework/8/src/main/java/asish/kdu/Spring_Exercise_3/repository/VehicleRepository.java:20]: in all functions in repo

File: backend/homework/8/src/main/java/asish/kdu/Spring_Exercise_3/repository/VehicleRepository.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+        db.remove(index);
+    }
+
+    public Vehicle findVehicle(String name) {
+        for(Vehicle v : db) {
+            if(v.name().equals(name)) return v;
+        }
+
+        return null;
```
Comment (rishav-kd) [inline backend/homework/8/src/main/java/asish/kdu/Spring_Exercise_3/repository/VehicleRepository.java:50]: dont return `null` throw an exception

**Review by rishav-kd — COMMENTED** [review]
looks good, try to focus on validations more. Try to utilise proper annotations for DTOs and handle edge scenarios in your code.

**Pull request closed without merging**